File: src/com/wix/rt/projectView/RTMoveProvider.java
/*
* Copyright 2000-2009 JetBrains s.r.o.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

package com.wix.rt.projectView;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.*;
import com.intellij.refactoring.move.MoveHandlerDelegate;
import com.intellij.refactoring.move.moveFilesOrDirectories.MoveFilesOrDirectoriesHandler;
import com.wix.rt.build.RTFileType;
import com.wix.rt.build.RTFileUtil;
import org.jetbrains.annotations.Nullable;

import java.util.Set;

/**
 * @author yole
 */
public class RTMoveProvider extends MoveHandlerDelegate {
    private static final Logger LOG = Logger.getInstance("#com.wix.rt.projectView.RTMoveProvider");

    @Override
    public boolean canMove(DataContext dataContext) {
        RTFile[] rtFiles = RTFile.DATA_KEY.getData(dataContext);
        return rtFiles != null && rtFiles.length > 0;
    }

    @Override
    public boolean isValidTarget(PsiElement psiElement, PsiElement[] sources) {
        return MoveFilesOrDirectoriesHandler.isValidTarget(psiElement);
    }

    public boolean canMove(PsiElement[] elements, @Nullable final PsiElement targetContainer) {
        return false;
    }

    @Override
    public void collectFilesOrDirsFromContext(DataContext dataContext, Set<PsiElement> filesOrDirs) {
        RTFile[] rtFiles = RTFile.DATA_KEY.getData(dataContext);
        LOG.assertTrue(rtFiles != null);
        if (rtFiles.length == 0) return;
        PsiFile[] jsToMove = new PsiFile[rtFiles.length];
        PsiFile[] filesToMove = new PsiFile[rtFiles.length];
        PsiFile[] controllersToMove = new PsiFile[rtFiles.length];
        for (int i = 0; i < rtFiles.length; i++) {
            jsToMove[i] = rtFiles[i].getRTJSFile();
            if (jsToMove[i] != null) {
                filesOrDirs.add(jsToMove[i].getContainingFile());
            }
            filesToMove[i] = rtFiles[i].getRtFile();
            if (filesToMove[i] != null) {
                filesOrDirs.add(filesToMove[i]);
            }
            controllersToMove[i] = rtFiles[i].getController();
            if (filesToMove[i] != null) {
                filesOrDirs.add(controllersToMove[i]);
            }
        }
    }

    @Override
    public boolean isMoveRedundant(PsiElement source, PsiElement target) {
        if (source instanceof PsiFile && source.getParent().equals(target)) {
            final VirtualFile virtualFile = ((PsiFile) source).getVirtualFile();
            if (RTFileUtil.isRTFile(virtualFile)) {
                return true;
            }
        }
        return super.isMoveRedundant(source, target);
    }
}

File: src/com/wix/rtk/projectView/RTMoveProvider.kt
package com.wix.rtk.projectView

import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.diagnostic.Logger
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.refactoring.move.MoveHandlerDelegate
import com.intellij.refactoring.move.moveFilesOrDirectories.MoveFilesOrDirectoriesHandler
import com.wix.rt.build.RTFileUtil
import com.wix.rt.projectView.RTFile

/**
 * @author yole
 */
class RTMoveProvider : MoveHandlerDelegate() {

    override fun canMove(dataContext: DataContext?): Boolean = RTFile.DATA_KEY.getData(dataContext!!)?.size ?: 0 > 0

    override fun isValidTarget(psiElement: PsiElement?, sources: Array<PsiElement>?): Boolean = MoveFilesOrDirectoriesHandler.isValidTarget(psiElement)

    override fun canMove(elements: Array<PsiElement>, targetContainer: PsiElement?): Boolean = false

    override fun collectFilesOrDirsFromContext(dataContext: DataContext?, filesOrDirs: MutableSet<PsiElement>?) {
        val rtFiles = RTFile.DATA_KEY.getData(dataContext!!)
        LOG.assertTrue(rtFiles != null)
        if (rtFiles?.size == 0) return
        val jsToMove = arrayOfNulls<PsiFile>(rtFiles!!.size)
        val filesToMove = arrayOfNulls<PsiFile>(rtFiles.size)
        val controllersToMove = arrayOfNulls<PsiFile>(rtFiles.size)
        for (i in rtFiles.indices) {
            jsToMove[i] = rtFiles[i].rtjsFile
            if (jsToMove[i] != null) {
                filesOrDirs!!.add(jsToMove[i]?.containingFile!!)
            }
            filesToMove[i] = rtFiles[i].rtFile
            if (filesToMove[i] != null) {
                filesOrDirs!!.add(filesToMove[i]!!)
            }
            controllersToMove[i] = rtFiles[i].controller
            if (filesToMove[i] != null) {
                filesOrDirs!!.add(controllersToMove[i]!!)
            }
        }
    }

    override fun isMoveRedundant(source: PsiElement?, target: PsiElement?): Boolean = source is PsiFile && source.parent == target && RTFileUtil.isRTFile(source.virtualFile) || super.isMoveRedundant(source, target)

    companion object {
        private val LOG = Logger.getInstance("#com.wix.rt.projectView.RTMoveProvider")
    }
}
